Enable reading and editing from same Configuration object

captainsafia opened the pull request:
**Changes in this PR**
 - Adds `WebAssemblyHostConfiguration` object that implements `IConfiguration, IConfigurationRoot, IConfigurationBuilder`
- Adds E2E test to validate reading from modified configuration
- Updates logging E2E tests to store log configuration in JSON file

Some notes:
- There's no real concept of "building" a configuration here. The configuration is ever mutable not frozen. I'm open to re-working this design if it's not ideal.
- It's a bit nestled in there but I removed the logic that set the minimum log level to warning by default. It is now up to the user what minimum log level to set.

Addresses #20231

Commit:
Enable reading and editing from same Configuration object

javiercn commented on src/Components/test/E2ETest/Tests/WebAssemblyConfigurationTest.cs:
```diff
@@ -47,6 +47,22 @@ public void WebAssemblyConfiguration_Works()
 
+        [Fact]
+        public void WebAssemblyConfiguration_ReloadingWorks()
+        {
```
This is D O P E

rynowak commented on src/Components/WebAssembly/WebAssembly/src/Hosting/MutableConfiguration.cs:
```diff
@@ -0,0 +1,197 @@
+    /// IConfigurationRoot, and IConfigurationBuilder. It can be used to simulatneously build
+    /// and read from a configuration object.
+    /// </summary>
+    public class MutableConfiguration : IConfiguration, IConfigurationRoot, IConfigurationBuilder
```
We should pick a different name given that this is part of the public API surface. Suggestion `WebAssemblyHostConfigurationBuilder`

captainsafia replied:
Renamed this to `WebAssemblyHostConfiguration`.

javiercn reviewed (commented):
Looks great so far!
A couple of high level pieces of feedback:
* Is there a way we can avoid exposing the "MutableConfiguration" type? For example, forcing a cast to IConfiguration if you want to access it or providing an extension method on the builder that does the cast an returns an IConfiguration instance. The main reason is that I don't like the type name, so I think an extension method or explicit cast is better, but I would love to hear your opinions on it, and other people's.
* Are we tracking what things we need to change after this change goes in?
  * We talked about the auth pieces on the template.
  * Did we have to do something else on the Logging front of things or are we doing it here? (I went through the PR really quick).

rynowak commented on src/Components/WebAssembly/WebAssembly/src/Hosting/MutableConfiguration.cs:
```diff
@@ -0,0 +1,197 @@
+        /// </remarks>
+        public IConfigurationSection GetSection(string key) => new ConfigurationSection(this, key);
+
+        public IEnumerable<IConfigurationSection> GetChildren() => GetChildrenImplementation(null);
```
Does it make sense to create an `IConfigurationRoot` object like build does (hold on to it), and then always delegate these implementations to it?

captainsafia replied:
I had started off with this model but hit a snag at this point and ended up with the approach you see here. I can't remember what the particular snag was but I'll refactor to that model and see if I can recall it.

rynowak commented on src/Components/WebAssembly/WebAssembly/src/Hosting/MutableConfiguration.cs:
```diff
@@ -0,0 +1,197 @@
+                }
+                return null;
+            }
+            set {
```
Do we need the setter here?

I guess I imagined that users would use this object to:
- Read keys (based on current/initial state)
- Add additional providers

captainsafia replied:
I did imagine that users would be able to set values here. This configuration object is never frozen. IMO, this enables a lot of interesting scenarios for users so I am partial to keeping it.

rynowak replied:
The reason I bring this up is that it adds some subtlety to the model that that wouldn't be there without it.

FYI `Set(...)` is generally implemented in a non-threadsafe way across all providers: https://github.com/dotnet/runtime/blob/master/src/libraries/Microsoft.Extensions.Configuration/src/ConfigurationProvider.cs#L25

Dictionary is thread-safe for multiple readers, as long as no one is writing. 

--- 

I actually wasn't even aware that `Set(...)` existed until your PR.

rynowak commented on src/Components/WebAssembly/WebAssembly/src/Hosting/MutableConfiguration.cs:
```diff
@@ -0,0 +1,197 @@
+    /// </summary>
+    public class MutableConfiguration : IConfiguration, IConfigurationRoot, IConfigurationBuilder
+    {
+        private readonly IList<IConfigurationProvider> _providers = new List<IConfigurationProvider>();
```
minor nitpick - using a type like `IList<>` over `List<>` *for your implementation details like members* generally just has the effect of making your code slower and more complex. `List<>` has a richer API that `IList<>` and you're unlikely to change back in the other direction if you just build the habit of using `List<>`.

captainsafia replied:
Got it. I can change this.

captainsafia replied:
Fixed.

rynowak commented on src/Components/WebAssembly/WebAssembly/src/Hosting/MutableConfiguration.cs:
```diff
@@ -0,0 +1,197 @@
+        /// <returns>The configuration value.</returns>
+        public string this[string key] {
+            get {
+                lock (mutex)
```
If we're doing synchronization for this, it needs to be in all of the methods that read/write mutable state. So that would include most of the functionality here.

I think you could also make the argument that we don't need to support synchronization because there's no way to do multi-threading in wasm, and also because one shouldn't hold on to this object after startup.

captainsafia replied:
There's some future-proofing here. It operates under the assumption that we'll have multi-threading soon and this is one less codepath to have to add support for thread safety to when that point arrives. I can probably remove it from this PR to limit the scope of the changes.

captainsafia commented:
> Is there a way we can avoid exposing the "MutableConfiguration" type? For example, forcing a cast to IConfiguration if you want to access it or providing an extension method on the builder that does the cast an returns an IConfiguration instance. The main reason is that I don't like the type name, so I think an extension method or explicit cast is better, but I would love to hear your opinions on it, and other people's.

Yeah, the name is not the best. I think Ryan's recommendation of `WebAssemblyHostConfiguration` is fine. That's actually where I started at first but I didn't like that name because it didn't highlight the key differentiation for the class.

> Are we tracking what things we need to change after this change goes in?

> We talked about the auth pieces on the template.

Yes, that is tracked in another issue.

> Did we have to do something else on the Logging front of things or are we doing it here? (I went through the PR really quick).

I've made the necessary changes in the E2E test. We had discussing adding some default logging configurations to the app settings in the template but we were back and forth on this and I'm not sure were we landed. In any case, it's trivial enough to add in.

pranavkm commented on src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostConfiguration.cs:
```diff
@@ -0,0 +1,190 @@
+    /// IConfigurationRoot, and IConfigurationBuilder. It can be used to simulatneously build
+    /// and read from a configuration object.
+    /// </summary>
+    public class WebAssemblyHostConfiguration : IConfiguration, IConfigurationRoot, IConfigurationBuilder
```
Is it possible to turn this `WebAssemblyHostConfiguration : IConfigurationRoot` to a has-a relation? i.e. can this use an instance of `ConfigurationRoot` for it's implementation?

pranavkm commented on src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostConfiguration.cs:
```diff
@@ -0,0 +1,190 @@
+        /// Gets the immediate descendant configuration sub-sections.
+        /// </summary>
+        /// <returns>The configuration sub-sections.</returns>
+        private IEnumerable<IConfigurationSection> GetChildrenImplementation(string path)
```
`GetChildrenImplementation` isn't re-used and `path` is always null Any reason it couldn't be merged in to `GetChildren`?

captainsafia replied:
Done.

captainsafia commented on src/Components/WebAssembly/WebAssembly/src/Hosting/MutableConfiguration.cs:
```diff
@@ -0,0 +1,197 @@
+        /// <summary>
+        /// Returns the sources used to obtain configuration values.
+        /// </summary>
+        public IList<IConfigurationSource> Sources => _sources;
```
Shouldn't need to as the `_sources` object is already a `List`.

SteveSandersonMS replied:
I suspect @pranavkm's concern is around mutability. Should it be possible for external code to mutate the `Sources` list directly? If not, returning a copy prevents that.

pranavkm commented on src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostConfiguration.cs:
```diff
@@ -0,0 +1,187 @@
+        /// <summary>
+        /// Returns the sources used to obtain configuration values.
+        /// </summary>
+        public IList<IConfigurationSource> Sources => _sources;
```
My concern with returning these as-is is does it weird out if you mutate the list as you're reading from it?

```
var children = config.GetChildren();
foreach (var item in children)
{
   config.Sources.RemoveAt(0);
}
```

or some weird combination like that. Returning a copy would avoid these it.

captainsafia replied:
Got it. Thanks for explaining your concern.

Practically, modifying the `Sources` object isn't as big a foot-shot as mutating the `Providers` object. Returning a shallow copy for both properties seems like a sensible guard against these situations.

captainsafia commented on src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostConfiguration.cs:
```diff
@@ -0,0 +1,190 @@
+        /// requirements of the IConfigurationBuilder and is not populated by
+        /// <see cref="WebAssemblyHostConfiguration"/> with any meaningful info.
+        /// </summary>
+        public IDictionary<string, object> Properties { get; } = new Dictionary<string, object>();
```
Fixed.

captainsafia commented on src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostConfiguration.cs:
```diff
@@ -0,0 +1,190 @@
+            set {
+                if (_providers.Count == 0)
+                {
+                    throw new InvalidOperationException();
```
Added an exception message here.

captainsafia commented on src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostConfiguration.cs:
```diff
@@ -0,0 +1,190 @@
+        /// <returns>The configuration value.</returns>
+        public string this[string key] {
+            get {
+                for (var i = _providers.Count - 1; i >= 0; i--)
```
Added a comment here.

captainsafia commented:
> Does it make sense to create an IConfigurationRoot object like build does (hold on to it), and then always delegate these implementations to it?

The trouble I ran into with having a `ConfigurationRoot` property that was maintained within this config construct is handling reloading of providers when the sources list is changed.

If we want to keep a `ConfigurationRoot` property synced with the sources that the user adds to the builder, we would have to repeatedly build a new instance by calling `new ConfigurationRoot(_providers)` within the `Add` method. The constructor calls `Load` on the provider. As we add more sources, the a new ConfigurationRoot object will be constructed to match the growing list of providers.

````csharp
new ConfigurationRoot([provider1]);
new ConfigurationRoot([provider1, provider2]);
...
new ConfiguratioNRoot([provider1, ..., providerN]);
````

Everytime you create a new `ConfigurationRoot` to add a new provider, you end up invoked `Load` on the n-1 providers before it. There's a few consequences to this, a notable one being that StreamProviders are not at all supported.

Let me know if you see a way to make ConfigurationRoot as an internal property work in a way that I don't.

rynowak commented:
> Everytime you create a new ConfigurationRoot to add a new provider, you end up invoked Load on the n-1 providers before it. There's a few consequences to this, a notable one being that StreamProviders are not at all supported.

I'm not worried about the cost of doing it, because adding a config provider is a rare operation that happens 1-2 times as most during initialization. 

You could also make the loading of the config root lazy, but if it doesn't simplify thhings, well then it doesn't 😆

rynowak commented:
Answered my own question: https://github.com/dotnet/runtime/blob/master/src/libraries/Microsoft.Extensions.Configuration/src/StreamConfigurationProvider.cs#L40

rynowak commented on src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostConfiguration.cs:
```diff
@@ -0,0 +1,187 @@
+        /// <summary>
+        /// Returns the sources used to obtain configuration values.
+        /// </summary>
+        public IList<IConfigurationSource> Sources => _sources.ToList();
```
I didn't follow the full discussion between you and @pranavkm (I was making chili). It looks like y'all settled on shallow copy. But if you want to spice this even more, you might think about returning a readonly wrapper here.

```C#
public IList<IConfigurationSource> Sources => new ReadOnlyCollection<IConfigurationSource>(_sources.ToList());
```

Then if someone tries to read they will get an exception instead of a silent failure. 

I'd also suggest making this property an explicit interface implementation:

```C#
IList<IConfigurationSource> IConfigurationBuilder.Sources => new ReadOnlyCollection<IConfigurationSource>(_sources.ToList());
```

That makes it *hidden* in a VIP only area. 

---

The fact that this has to exist is an unfortunate smell. This interface is weird for having both a public mutable collection and an `Add` method.

captainsafia replied:
Done and done.

SteveSandersonMS commented on src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostConfiguration.cs:
```diff
@@ -0,0 +1,187 @@
+        public IList<IConfigurationSource> Sources => _sources.ToList();
+
+        /// <summary>
+        /// Returns the providers user to obtain configuration values.
```
```suggestion
        /// Gets the providers used to obtain configuration values.
```

captainsafia replied:
📝 🤦

SteveSandersonMS commented on src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostConfiguration.cs:
```diff
@@ -0,0 +1,187 @@
+        private ConfigurationReloadToken _changeToken = new ConfigurationReloadToken();
+
+        /// <summary>
+        /// Returns the sources used to obtain configuration values.
```
```suggestion
        /// Gets the sources used to obtain configuration values.
```

SteveSandersonMS commented on src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostConfiguration.cs:
```diff
@@ -0,0 +1,187 @@
+
+        /// <summary>
+        /// Gets a key/value collection that can be used to share data between the <see cref="IConfigurationBuilder"/>
+        /// and the registered <see cref="IConfigurationProvider"/>s.
```
```suggestion
        /// and the registered <see cref="IConfigurationProvider"/> instances.
```

SteveSandersonMS commented on src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostConfiguration.cs:
```diff
@@ -0,0 +1,187 @@
+        /// Gets a key/value collection that can be used to share data between the <see cref="IConfigurationBuilder"/>
+        /// and the registered <see cref="IConfigurationProvider"/>s.
+        ///
+        /// In this implementation, this largely exists as a way to satisfy the
+        /// requirements of the IConfigurationBuilder and is not populated by
+        /// <see cref="WebAssemblyHostConfiguration"/> with any meaningful info.
```
Since this is on a public type, I suggest reducing these lines to regular `//` comments, outside the XML doc comment block. Customers probably don't need to know this since it's about an implementation detail.

davidfowl commented on src/Components/WebAssembly/WebAssembly/test/Hosting/WebAssemblyHostBuilderTest.cs:
```diff
@@ -204,7 +204,7 @@ public void Build_AddsConfigurationToServices()
 
             // Assert
-            var configuration = host.Services.GetRequiredService<IConfiguration>();
+            var configuration = host.Services.GetRequiredService<WebAssemblyHostConfiguration>();
```
Why this change?

captainsafia replied:
I changed the DI to inject the `WebAssemblyHostConfiguration` so that users could access the type with the read/write enabled. I'll probably change the DI to register both the `IConfiguration` and `WebAssemblyHostConfiguration` type to have backwards compatibility and expose the new type to users.

davidfowl replied:
Do we need to expose the WebAssemblyHostConfiguration  at all?

captainsafia replied:
![image](https://user-images.githubusercontent.com/1857993/78931442-cf319580-7a5a-11ea-8792-ff849d4c37d7.png)

After noodling on it, nope.

SteveSandersonMS commented on src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostConfiguration.cs:
```diff
@@ -0,0 +1,187 @@
+
+                foreach (var provider in _providers)
+                {
+                    provider.Set(key, value);
```
I'm not an expert on the configuration APIs, but find this a bit surprising. Do we definitely want to set the value on *all* the providers? Would it suffice to have a single internal "mutable provider" which is where we write things that get set this way? Then the `_providers.Count == 0` check could be dropped too.

javiercn replied:
Do we expect this to be the common case? I somewhat agree with Steve here. I would keep an instance of the last in memory provider and reuse that if possible instead of setting all the providers. What if a provider is "read-only"?

javiercn replied:
Nevermind, this does the same thing as https://github.com/dotnet/runtime/blob/master/src/libraries/Microsoft.Extensions.Configuration/src/ConfigurationRoot.cs#L52-L80

captainsafia replied:
Yes, it's doing the same thing as the ConfigurationRoot.

I did some GitHub archaeology and this code [was introduced](https://github.com/aspnet/Configuration/commit/8edcceec69304365d8392fef85ebe68854ef13f0) about five years ago in the configuration provider.

I had assumed that the goal was to ensure that all providers were in sync with each other. So if I set a value using `Configuration["test"]`, I would expect all child providers to set that value.

Also, I think having the case that errors if no providers are registered is helpful. Maintaining an internal `MemoryProvider` seems like it would be surprising to the end-user.

SteveSandersonMS replied:
OK, thanks for figuring that out!

davidfowl commented on src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostBuilder.cs:
```diff
@@ -52,13 +52,11 @@ internal WebAssemblyHostBuilder(WebAssemblyJSRuntimeInvoker jsRuntimeInvoker)
+            Configuration = new WebAssemblyHostConfiguration();
             RootComponents = new RootComponentMappingCollection();
             Services = new ServiceCollection();
             Logging = new LoggingBuilder(Services);
```
Should we be calling `Logging.AddConfiguration(Configuration)`?

captainsafia replied:
Yes, we are planning on doing this in a follow-up PR.

SteveSandersonMS commented on src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostConfiguration.cs:
```diff
@@ -0,0 +1,187 @@
+                throw new ArgumentNullException(nameof(source));
+            }
+
+            // Ad this source and its associated provided to the source
```
```suggestion
            // Add this source and its associated provider to the source
```

SteveSandersonMS commented on src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostBuilder.cs:
```diff
@@ -177,16 +175,15 @@ public void ConfigureContainer<TBuilder>(IServiceProviderFactory<TBuilder> facto
             // Intentionally overwrite configuration with the one we're creating.
-            var configuration = Configuration.Build();
-            Services.AddSingleton<IConfiguration>(configuration);
+            Services.AddSingleton<WebAssemblyHostConfiguration>(Configuration);
```
I think it needs to stay as `Services.AddSingleton<IConfiguration>(configuration)` so that consumers can still receive instances from DI using `@inject IConfiguration MyConfig`.

If we also wanted people to be able to receive it via `@inject WebAssemblyHostConfiguration MyConfig` we would need a second registration, but I don't think there's any use case for that.

SteveSandersonMS replied:
It is a bit quirky that ASP.NET Core's DI system doesn't resolve requests for subtypes, but does simplify customers' understanding of what's going on (no need for prioritisation rules).

captainsafia replied:
> It is a bit quirky that ASP.NET Core's DI system doesn't resolve requests for subtypes, but does simplify customers' understanding of what's going on (no need for prioritisation rules).

True. I've opted to remove it from the dependency injection. Users still have the flexibility to register it themselves within their startup logic but I think it is sensible to not auto-do it here.

SteveSandersonMS commented:
@captainsafia This is looking really great!

pranavkm commented on src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostConfiguration.cs:
```diff
@@ -0,0 +1,187 @@
+        /// <summary>
+        /// Returns the providers user to obtain configuration values.
+        /// </summary>
+        public IEnumerable<IConfigurationProvider> Providers => _providers.ToList();
```
If we could make these explicit implementations, that seems great. It's not very useful for users to interact with these properties.

pranavkm commented on src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostConfiguration.cs:
```diff
@@ -0,0 +1,187 @@
+        /// <inheritdoc />
+        public string this[string key]
+        {
+            get {
```
Braces on the next line

pranavkm commented on src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostConfiguration.cs:
```diff
@@ -0,0 +1,187 @@
+
+                return null;
+            }
+            set {
```
Braces

pranavkm commented on src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostConfiguration.cs:
```diff
@@ -0,0 +1,187 @@
+        /// Gets the immediate descendant configuration sub-sections.
+        /// </summary>
+        /// <returns>The configuration sub-sections.</returns>
+        public IEnumerable<IConfigurationSection> GetChildren()
```
I'm not super sure how commonly this API is used. Perhaps this could be an explicit implementation? @anurse thoughts \ suggestions?

anurse replied:
What's the context? The API `IConfiguration.GetChildren` is certainly used by some components.

javiercn replied:
Yeah, I've used this API in the past, the implementation for that should be public

captainsafia replied:
IMO, I actually think this should be an explicit implementation. I don't see a scenario where the user would want to call `GetChildren()` directly on `WebAssemblyHostConfiguration`.  It might also be sensible to start with a more restrictive API then open it up if that is something users express interested in it.

anurse replied:
I guess I'm not clear why you'd want to **hide** this. It's something a user would *expect* to see when accessing an implementation of `IConfiguration` or something that looks like one.

Does it not work as expected in this scenario? I'm not versed in the context for the change at large though.

captainsafia replied:
This change introduces a new class that implements both the IConfiguration and IConfigurationBuilder interfaces. Making it an explicit implementation requires that the user be accessing the object as an `IConfiguartion` which is the behavior that we want to encourage.

anurse replied:
Ah, ok. That makes a lot of sense then :). You'd probably also want to do it for things like `GetSection` and the indexer then.

pranavkm commented on src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostConfiguration.cs:
```diff
@@ -0,0 +1,187 @@
+        public IEnumerable<IConfigurationSection> GetChildren()
+        {
+            return Providers
+                .Aggregate(Enumerable.Empty<string>(),
```
You might be able to use `SelectMany` here instead of `Aggregate`

`Providers.SelectMany(s => source.GetChildKeys(Enumerable.Empty<string>(), null))`

pranavkm commented on src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostConfiguration.cs:
```diff
@@ -0,0 +1,187 @@
+                .Aggregate(Enumerable.Empty<string>(),
+                    (seed, source) => source.GetChildKeys(seed, null))
+                .Distinct(StringComparer.OrdinalIgnoreCase)
+                .Select(key => this.GetSection(key));
```
Consider a `ToList` here. Same sort of issue with `providers`, you could have issues if `Providers` changed while this was being iterated on.

pranavkm reviewed (commented):
Nearly there.

captainsafia commented:
Thanks everyone for the review! I've addressed the actionable items that folks pointed out above.

A few folks asked why `IConfigurationRoot` was implemented as an interface and not maintained as an internal property. I explained some of the reasons I chose not to do this [here](https://github.com/dotnet/aspnetcore/pull/20647#issuecomment-611271608). Generally, I wanted to have more fine-grained control over when providers were reloaded to avoid:

- Bugs that occur when providers that can only be loaded once are reloaded
- I think the behavior that adding a provider will reload all existing providers is a little unexpected and surprising to users.

This choice does have a fair bit of trade-offs, especially with regard to all the interfaces that need to be implemented instead of delegated to the builder property. But IMO, it's worth it for more explicit control over the providers list.

Also, there was some general feedback about whether or not WebAssemblyHostConfiguration should be registered in the DI. After some 🤔, I've removed it.

javiercn commented on src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostConfiguration.cs:
```diff
@@ -0,0 +1,189 @@
+                if (_providers.Count == 0)
+                {
+                    throw new InvalidOperationException("Can only set property if at least one provider has been inserted.");
+                }
```
Should we have a default in memory provider to make this default case work?
Does the current implementaiton mean that just `WebAssemblyHostBuilder.Configuration["property"] = "something"` won't work and that you would need to do something like `WebAssemblyHostBuilder.Configuration.AddProvider(new InMemoryProvider(new Dictionary<string,string>{ ["property"] = "something})` instead?

javiercn replied:
Same thing here.

javiercn commented on src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostConfiguration.cs:
```diff
@@ -0,0 +1,189 @@
+        /// <summary>
+        /// Gets the providers used to obtain configuration values.
+        /// </summary>
+        IEnumerable<IConfigurationProvider> IConfigurationRoot.Providers => new ReadOnlyCollection<IConfigurationProvider>(_providers.ToList());
```
👍

javiercn commented on src/Components/WebAssembly/WebAssembly/test/Hosting/WebAssemblyHostConfigurationTest.cs:
```diff
@@ -0,0 +1,111 @@
+            // Assert
+            Assert.NotNull(children);
+            Assert.NotEmpty(children);
+        }
```
Should we have a test with two or more providers here to ensure that the results are what we expect? Using only one memory provider only covers the simplest case.

Could I suggest we also test a config that is a bit more complex? (like with nested sections) to ensure that the expected override behavior is preserved.

captainsafia replied:
I copied a few more of the nested test cases from [this](https://github.com/dotnet/runtime/blob/9e608fcecbabbc651d21582416e5807038ca7142/src/libraries/Microsoft.Extensions.Configuration/tests/ConfigurationTest.cs).

I added a test with an in-memory provider and a custom one. The E2E tests validate with a JsonStreamProvider and a MemoryProvider.

javiercn reviewed (approved):
Looks fantastic!

I had a couple of small comments about some improvements to the tests that I'll let you handle.

SteveSandersonMS reviewed (approved):
This looks awesome now!

captainsafia commented:
Thanks all for the reviews. I'll go ahead and merge this now. There is [one comment](https://github.com/dotnet/aspnetcore/pull/20647#discussion_r406536988) that I will address in a follow-on.
